tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/SetInRuleWithArrays.scala

```diff
@@ -0,0 +1,83 @@
+package at.forsyte.apalache.tla.bmcmt.rules
+
+import at.forsyte.apalache.tla.bmcmt.implicitConversions.Crossable
+import at.forsyte.apalache.tla.bmcmt.rewriter.ConstSimplifierForSmt
+import at.forsyte.apalache.tla.bmcmt.types.{BoolT, CellT, FinSetT, PowSetT}
+import at.forsyte.apalache.tla.bmcmt.{ArenaCell, SymbState, SymbStateRewriter, types}
+import at.forsyte.apalache.tla.lir.TlaEx
+import at.forsyte.apalache.tla.lir.convenience.tla
+import at.forsyte.apalache.tla.lir.UntypedPredefs._
+
+class SetInRuleWithArrays(rewriter: SymbStateRewriter) extends SetInRule(rewriter) {
+  private val simplifier = new ConstSimplifierForSmt
+
+  override protected def powSetIn(state: SymbState, powsetCell: ArenaCell, elemCell: ArenaCell): SymbState = {
+    def checkType: PartialFunction[(CellT, CellT), Unit] = {
+      case (PowSetT(FinSetT(expectedType)), FinSetT(actualType)) =>
+        assert(expectedType == actualType)
+    }
+    // double check that the type finder did its job
+    checkType(powsetCell.cellType, elemCell.cellType)
+
+    val leftElems = state.arena.getHas(elemCell)
+    val rightDomain = state.arena.getDom(powsetCell)
+    val rightDomainElems = state.arena.getHas(rightDomain)
+    // EqConstraints need to be generated, since missing in-relations, e.g. in sets of tuples, will lead to errors.
+    // TODO: Inlining this method is pointless. We should consider handling tuples and other structures natively in SMT.
```

Comment:
Good point. We should figure out how to deal with records that can be used as discriminated unions. Perhaps, ADTs should help us.